### ISSUE TITLE
Semaphore blocked when fetching multiple Storage S3 files

### ISSUE THREAD

**Issue by SvyatHoly** [aws-amplify/amplify-ios#1418]
### Describe the bug

Amplify can't fetchAuthSession from multiple queues, so all tasks gonna freeze on semaphore here:

```
public func getIdentityId() -> Result<String, AuthError> {
        var result: Result<String, AuthError>?
        let semaphore = DispatchSemaphore(value: 0)
        _ = Amplify.Auth.fetchAuthSession { event in
            defer {
                semaphore.signal()
            }

            switch event {
            case .success(let session):
                result = (session as? AuthCognitoIdentityProvider)?.getIdentityId()
            case .failure(let error):
                result = .failure(error)

            }
        }
        semaphore.wait()
        guard let validResult = result else {
            return .failure(AuthError.unknown("""
            Did not receive a valid response from fetchAuthSession for identityId.
            """))
        }
        return validResult
    }
```

<img width="728" alt="Screenshot 2021-09-11 at 02 46 52" src="https://user-images.githubusercontent.com/43349266/132928772-68c16883-b8ea-46b8-aed5-58f98bf3f127.png">

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Create multiple concurrent queue:
    private static let urlSessionQueue: OperationQueue = {
        let queue = OperationQueue()
        queue.maxConcurrentOperationCount = 1
        return queue
    }()
2. try do execute multiple tasks with Storage:
urlSessionQueue.addOperation {
    Amplify.Storage.downloadData(key: ...)
3. Nothing happens
```

### Expected behavior

Data can be fetched from multiple queues.

### Amplify Framework Version

1.12.0

### Amplify Categories

Storage

### Dependency manager

Swift PM

### Swift version

5.0

### CLI version

4.52.0

### Xcode version

Version 12.5.1

### Device

Iphone 11

### iOS Version

IOS 14.7.1

**Comment by anatomybook** [comment]
I'm seeing to the same issue. I've bypassed it by executing all my S3 interaction with a serial queue. However, I'd love to see this fixed as well.

**Comment by royjit** [comment]
Related issue - https://github.com/aws-amplify/amplify-ios/issues/443

**Comment by brennanMKE** [comment]
@SvyatHoly What is happening here is many work items are being started on a concurrent queue while calling the `getIdentityId` function which is waiting on a semaphore while the work to refresh access is running. What happens is the scheduler will attempt to run all work on the concurrent queue and if there are no threads which are not blocked it will try to create a new thread. At some point the system will not allow your process to get another thread which will cause your app to stall and possibly deadlock. Please see the section called *Avoiding Excessive Thread Creation* on the [DispatchQueue](https://developer.apple.com/documentation/dispatch/dispatchqueue) doc which explains this common problem. 

There is another issue which is a good reference. See #1380. In this issue it was calling [waitUntilFinished](https://developer.apple.com/documentation/foundation/nsoperation/1409256-waituntilfinished) which blocks the current thread. What you want to do is use the non-blocking API so that work on the concurrent queue is not blocked which forces new threads to be created. What you will have instead is closures which will be run once the work done for identity has completed. It won't require new threads even if you initiate many new operations.

Apple often encourages the use of non-blocking APIs in WWDC talks. You can find these session videos via [this search](https://www.google.com/search?client=safari&rls=en&q=non-blocking+site%3Adeveloper.apple.com%2Fvideos%2Fplay%2F&ie=UTF-8&oe=UTF-8) One recent one covers the cause of hangs and how to eliminate them. See [How to eliminate hangs](https://developer.apple.com/videos/play/wwdc2021/10258/?time=868).

We will be offering a non-blocking API to replace the code you are using which 

```swift
// Blocking API
public func getIdentityId() -> Result<String, AuthError> {}

// Non-Blocking API
public func getIdentityId(completionHandler: @escaping (Result< String, Error>) -> Void) {}
```

Note it does not have a return type and instead provides an escaping closure which will pass back the `Result`.

In the meantime, you could use the code below to create a Non-Blocking API which calls `getIdentityId` and then calls the completion handler with the `Result`.

```swift
struct AuthWrapper {
    let identityProvider: AuthCognitoIdentityProvider

    init(identityProvider: AuthCognitoIdentityProvider) {
        self.identityProvider = identityProvider
    }

    func getUserTokens(completionHandler: @escaping (Result<String, Error>) -> Void) {
        do {
            let tokens = try identityProvider.getIdentityId().get()
            completionHandler(.success(tokens))
        } catch {
            completionHandler(.failure(error))
        }
    }

}
```

**Comment by brennanMKE** [comment]
Please try this solution and let us know how it works for you.

**Comment by brennanMKE** [comment]
@SvyatHoly There is a [related issue](https://github.com/aws-amplify/amplify-ios/issues/1279#issuecomment-963422363) and I've worked out a solution there which you may be able to use.

**Comment by brennanMKE** [comment]
[#3872](https://github.com/aws-amplify/aws-sdk-ios/pull/3872)

This update will be included in the next release which will eliminate the thread locking you see in these stack frames.
